onnxruntime/core/providers/cuda/nn/dropout_impl.cu

```diff
@@ -44,26 +44,60 @@ __global__ void DropoutKernel(
   curandStatePhilox4_32_10_t state;
   curand_init(seeds.first, idx, seeds.second, &state);
 
-  // We ensure every thread generates the same number of random numbers (by rounding
-  // up the size) and at the same timestep (by syncing threads).
-  // From CUDA curand documentation:
-  //   The Philox_4x32_10 algorithm is closely tied to the thread and block count.
-  //   Each thread computes 4 random numbers in the same time thus the most efficient
-  //   use of Philox_4x32_10 is to generate a multiple of 4 times number of threads.
-  for (CUDA_LONG id = idx; id < rounded_size; id += step_size) {
-    float4 rand = curand_uniform4(&state);
-  
-  #pragma unroll
-    for (CUDA_LONG i = 0; i < UNROLL; i++) {
-      CUDA_LONG li = id + gridDim.x * blockDim.x * i;
-      if (li < N) {
-        mask_data[li] = (&rand.x)[i] < p;
-        Y_data[li] = T(float(X_data[li]) * mask_data[li] * scale);
+  float4 rand;
+
+  if (N % UNROLL != 0) {
+    // We ensure every thread generates the same number of random numbers (by rounding
+    // up the size) and at the same timestep (by syncing threads).
+    // From CUDA curand documentation:
+    //   The Philox_4x32_10 algorithm is closely tied to the thread and block count.
+    //   Each thread computes 4 random numbers in the same time thus the most efficient
+    //   use of Philox_4x32_10 is to generate a multiple of 4 times number of threads.
+    for (CUDA_LONG id = idx; id < rounded_size; id += step_size) {
+      rand = curand_uniform4(&state);
+    
+      #pragma unroll
+      for (CUDA_LONG i = 0; i < UNROLL; i++) {
+        CUDA_LONG li = id + gridDim.x * blockDim.x * i;
```

Comment:
as Weixing suggested, vectorized and non-vectorized kernels are using different thread-to-id mapping... 
vectorized kernel are handling 4 contiguous elements per thread, non-vectorized kernels are handling 4 elements "step_size" apart... 
We should simplify this...

Reply:
Same comments applied to Dropout, DropoutGrad and BiasDropout kernels.

Reply:
If no perf improvement, we can leave the old access pattern as is. 
@weixingzhang